### PR TITLE
Added rent dates validation

### PR DESCRIPTION
**Pull request by storresy**
**Summary**
Added custom validation to rent model, that validation works in this way:
- Do nothing if book, start date or end date are missing
- Search for a rent for the same book that overlays with the selected dates
- Add an error if a rent was found in the previous step

**Trello Card**
https://trello.com/c/q4EYfTg0

**Screenshots**
screen shot 2018-12-19 at 15 44 32
screen shot 2018-12-19 at 14 42 16
screen shot 2018-12-19 at 14 42 45
screen shot 2018-12-19 at 14 44 06
screen shot 2018-12-19 at 14 46 28

### CODE REVIEW

File: app/models/rent.rb [hunk 1]
```diff
@@ -3,7 +3,31 @@ class Rent < ApplicationRecord
...
   validates :user_id, :book_id, :start_date, :end_date, presence: true
+  validate :book_availability
 
   scope :rented_by, ->(user_id) { where(user_id: user_id) }
   scope :ending_today, -> { where(end_date: Time.zone.today) }
+
+  private
+
+  def book_availability
+    return if book_id.blank? || start_date.blank? || end_date.blank?
```
Comment (Gmarn) [inline app/models/rent.rb:14]: All of this check are validates by `presence: true` just above, aren't they?

File: app/models/rent.rb [hunk 2]
```diff
@@ -3,7 +3,31 @@ class Rent < ApplicationRecord
...
+                                                         condition_params)
+    overlaping_rent = overlaping_rent.where.not(id: id) if id.present?
+    overlaping_rent.exists?
+  end
+
+  def book_availability_conditions
+    'start_date <= :start_date AND end_date >= :start_date
+     OR start_date <= :end_date AND end_date >= :end_date
+     OR start_date >= :start_date AND start_date <= :end_date
+     OR end_date >= :start_date AND end_date <= :end_date'
```
Comment (Gmarn) [inline app/models/rent.rb]: This query could be simplify a lot...
With only 2 conditions.

File: app/models/rent.rb [hunk 3]
```diff
@@ -3,7 +3,31 @@ class Rent < ApplicationRecord
...
+
+  def book_availability
+    return if book_id.blank? || start_date.blank? || end_date.blank?
+
+    errors.add(:book_id, 'is not available in that timeband') if overlaping_rent_exists?
+  end
+
+  def overlaping_rent_exists?
+    condition_params = { end_date: end_date, start_date: start_date }
+    overlaping_rent = Rent.where(book_id: book_id).where(book_availability_conditions,
```
Comment (Gmarn) [inline app/models/rent.rb]: Use your current model: `book.rents` to find all associate rents.

**Review by Gmarn — CHANGES_REQUESTED** [review]
I am sure you can refactor this and did it in 2 or 4 lines ;)

File: app/models/rent.rb (hunk 1 above)
Comment (storresy) [reply app/models/rent.rb:14]: If I remove that line I could get an error when querying for a rent that overlaps with the selected time frame

File: app/models/rent.rb (hunk 2 above)
Comment (storresy) [reply app/models/rent.rb]: Thank you. I think I learnt a lot trying to solve this requested change.

**Review by Gmarn — APPROVED** [review]

**Pull request merged**